Make some of the existing rules optional and add...

diego-ojeda-binbash opened the pull request:
...a couple of rules that can be used for HIPAA compliance.

dynamike commented on variables.tf:
```diff
@@ -126,3 +126,43 @@ variable "required_tags" {
+  default     = false
+}
+
+variable "check_rds_storage_encrypted" {
```
I see you defined the flag, but it doesn't appear to be getting used anywhere.

exequielrafaela replied:
Adding `count` in order to allow this rule as optional. 

```terraform
resource "aws_config_config_rule" "rds-storage-encrypted" {
  count        = var.check_rds_storage_encrypted ? 1 : 0
  name        = "rds-storage-encrypted"
  description = "Checks whether storage encryption is enabled for your RDS DB instances."

  source {
    owner             = "AWS"
    source_identifier = "RDS_STORAGE_ENCRYPTED"
  }

  depends_on = [aws_config_configuration_recorder.main]
}
```

dynamike commented on config-rules.tf:
```diff
@@ -295,6 +299,51 @@ resource "aws_config_config_rule" "required-tags" {
+  depends_on = [aws_config_configuration_recorder.main]
+}
+
+#==================================================
```
What is 0.3 referring too?

exequielrafaela replied:
We'll clean up the unnecessary comments here.

dynamike commented on config-rules.tf:
```diff
@@ -295,6 +299,51 @@ resource "aws_config_config_rule" "required-tags" {
+}
+
+#==================================================
+# 0.1	Checks whether running instances are using specified AMIs.
```
What is 0.1 referring too?

exequielrafaela replied:
We'll clean up the unnecessary comments here.

dynamike commented on variables.tf:
```diff
@@ -126,3 +126,43 @@ variable "required_tags" {
+}
+
+variable "ami_required_tag_key_value" {
+  description = "Tag key and value which AMI has to have to be compliant: Example: tag-key:tag-value"
```
Since you can add multiple tags, can you mention this is a comma separated e.g., `key1:value1,key2:value2`?

exequielrafaela replied:
We'll be updated as follows 

```terraform
variable "ami_required_tag_key_value" {
  description = "Tag/s key and value which AMI has to have in order to be compliant: Example: key1:value1,key2:value2"
  type        = "string"
  default     = ""
}
```

dynamike commented on variables.tf:
```diff
@@ -126,3 +126,43 @@ variable "required_tags" {
+
+variable "check_rds_storage_encrypted" {
+  description = "Enable rds-storage-encrypted"
+  default     = false
```
The default for this should probably be true

exequielrafaela replied:
Agreed :handshake:

dynamike commented on variables.tf:
```diff
@@ -126,3 +126,43 @@ variable "required_tags" {
+
+variable "check_ec2_encrypted_volumes" {
+  description = "Enable ec2-encrypted-volumes"
+  default     = false
```
This is a sane default in my opinion and can be default enabled.

exequielrafaela replied:
I strongly agree here too :+1:

dynamike commented:
One more thing. Could you update the new checks you added unders `The following AWS Config Rules are supported:` in the README?

dynamike commented on config-rules.tf:
```diff
@@ -295,6 +299,51 @@ resource "aws_config_config_rule" "required-tags" {
+    source_identifier   = "ENCRYPTED_VOLUMES"
+  }
 
   depends_on = [aws_config_configuration_recorder.main]
```
It looks like you've removed `depends_on = [aws_config_configuration_recorder.main]` from the `required-tags` config rule.

exequielrafaela replied:
Great catch here, `depends_on` back in place, sorry about that.

```terraform
resource "aws_config_config_rule" "required-tags" {
  count = var.check_required_tags ? 1 : 0

  name        = "required-tags"
  description = "Checks if resources are deployed with configured tags."

  scope {
    compliance_resource_types = var.required_tags_resource_types
  }

  input_parameters = "${jsonencode(var.required_tags)}"

  source {
    owner             = "AWS"
    source_identifier = "REQUIRED_TAGS"
  }

  depends_on = [aws_config_configuration_recorder.main]
}
```

diego-ojeda-binbash commented:
> Thanks for the PR. This is a great stuff. I have a couple of comments.
> 
> pre-commit checks are failing, be sure to run
> 
> ```
> pre-commit install --install-hooks
> pre-commit run --all-files
> ```

Hi @dynamike , thanks for all your feedback! I probably rushed to open this PR. Either I or somebody from my team will be addressing all your comments soon. 👍

dynamike commented:
any update on this?

exequielrafaela commented:
@dynamike I'll collaborate here during this next few days. Thanks for the follow up.

exequielrafaela commented:
> Thanks for the PR. This is a great stuff. I have a couple of comments.
> 
> pre-commit checks are failing, be sure to run
> 
> ```
> pre-commit install --install-hooks
> pre-commit run --all-files
> ```

 - pre-commit reqs passing now

```
╭─delivery at delivery-I7567 in ~/Binbash/repos/BB-Leverage/terraform/terraform-aws-config on BBL-153-sync-pr-updates✘✘✘ using ‹› 19-12-05 - 13:33:01
╰─○ make pre_commit_tests
pre-commit run --all-files
Check JSON...........................................(no files to check)Skipped
Check for merge conflicts................................................Passed
Check Yaml...............................................................Passed
Detect Private Key.......................................................Passed
Pretty format JSON...................................(no files to check)Skipped
Trim Trailing Whitespace.................................................Passed
markdownlint.............................................................Passed
Terraform docs...........................................................Passed
Terraform fmt............................................................Passed
golangci-lint............................................................Passed
hookid: golangci-lint

level=info msg="[config_reader] Config search paths: [./ /home/delivery/Binbash/repos/BB-Leverage/terraform/terraform-aws-config /home/delivery/Binbash/repos/BB-Leverage/terraform /home/delivery/Binbash/repos/BB-Leverage /home/delivery/Binbash/repos /home/delivery/Binbash /home/delivery /home /]"
level=info msg="[config_reader] Used config file .golangci.yml"
level=info msg="[lintersdb] Active 14 linters: [deadcode errcheck gofmt goimports golint gosec gosimple govet ineffassign staticcheck structcheck typecheck unused varcheck]"
level=info msg="[loader] Go packages loading at mode 575 (deps|files|imports|types_sizes|compiled_files|exports_file|name) took 299.606475ms"
level=info msg="[runner/filename_unadjuster] Pre-built 0 adjustments in 174.341µs"
level=info msg="[runner/goanalysis_metalinter/goanalysis] analyzers took 0s with no stages"
level=info msg="[runner/unused/goanalysis] analyzers took 0s with no stages"
level=info msg="[runner] processing took 3.412µs with stages: max_same_issues: 854ns, filename_unadjuster: 305ns, skip_dirs: 292ns, nolint: 280ns, max_from_linter: 239ns, exclude: 173ns, cgo: 154ns, diff: 151ns, path_shortener: 140ns, identifier_marker: 139ns, skip_files: 137ns, path_prettifier: 130ns, source_code: 128ns, autogenerated_exclude: 120ns, max_per_file_from_linter: 59ns, uniq_by_line: 57ns, exclude-rules: 54ns"
level=info msg="[runner] linters took 137.792977ms with stages: goanalysis_metalinter: 137.44297ms, unused: 300.159µs"
level=info msg="File cache stats: 0 entries of total size 0B"
level=info msg="Memory: 6 samples, avg is 69.0MB, max is 69.4MB"
level=info msg="Execution took 441.776883ms"
```

I'll run the E2E terratests and when everything looks good :heavy_check_mark:  proceed to commit the changes :checkered_flag:  
@dynamike we'll keep you posted! :handshake:  

CC: @diego-ojeda-binbash

exequielrafaela commented:
@dynamike thanks for your patient here, every requested change comment has been addressed. Please let us know if there is anything else that needs to be reviewed from our side prior to the definite approval and merge.

We look forward to your feedback and comments.

CC: @diego-ojeda-binbash @mpagnucco @gdmlnx
